### PR TITLE
Fixes for SCM Model

### PR DESCRIPTION
**Pull request by mathomp4**
This is a work-in-progress PR to fix the SCM. Note: This does not work yet. It is mainly a branch that incorporates changes from @narnold1 and myself from https://github.com/GEOS-ESM/GEOSgcm_GridComp/pull/604 but brought up to present day GEOSgcm_GridComp.

The changes are:

1. Add `VARFLT` import and `TROPK_BLENDED` export to datmo → NOTE: This was done *very* simply, they are not filled
2. Reorder the setup routines in Moist so that microphysics setup runs first because binary restarts expect `Q` first
3. Pass in an MPI communicator to `gfdl_cloud_microphys_init` because in SCM, this is the first call to `fms_init()` and it was crashing without a comm (usually FV3 initialize FMS)

---

UPDATE! It looks like this PR in concert with https://github.com/GEOS-ESM/GEOSgcm_App/pull/477 works. Huzzah!

As such I'll undraft and remote the WIP.

### CODE REVIEW

**Comment by mathomp4** [comment]
With the current changes, I can get past any issues in Moist, but then you die with:

```
 Character Resource Parameter: VEGDYN_INTERNAL_RESTART_FILE:vegdyn.data
 NOT using buffer I/O for file: vegdyn.data
pe=00000 FAIL at line=00602    BinIO.F90                                <status=5001>
pe=00000 FAIL at line=00453    BinIO.F90                                <status=5001>
pe=00000 FAIL at line=06186    MAPL_Generic.F90                         <status=5001>
 ESMF_StatePrint: (pet 0):
  State name: VEGDYN_INTERNAL
     status: Unspecified intent direction, object count: 3
     reconcile needed: T
 Base name    = VEGDYN_INTERNAL
 Status: Base = Ready,  object = Ready
 Proxy        = no
 Root Info (Attributes):
{
  "ESMF": {
    "General": {
      "MAPL_GridTypeBits": 2,
      "MAPL_RestartRequired": 0,
      "MAPL_StateItemOrderList": [
        "ITY",
        "Z2CH",
        "ASCATZ0"
      ],
      "POSITIVE": "down"
    }
  }
}     object: 1,name: ASCATZ0
            type: Field
     object: 2,name: ITY
            type: Field
     object: 3,name: Z2CH
            type: Field
pe=00000 FAIL at line=01614    MAPL_Generic.F90                         <unknown error>
pe=00000 FAIL at line=01051    MAPL_Generic.F90                         <status=-1>
pe=00000 FAIL at line=01818    MAPL_Generic.F90                         <status=-1>
pe=00000 FAIL at line=01272    MAPL_Generic.F90                         <status=-1>
pe=00000 FAIL at line=01044    MAPL_Generic.F90                         <status=-1>
pe=00000 FAIL at line=01818    MAPL_Generic.F90                         <status=-1>
pe=00000 FAIL at line=01272    MAPL_Generic.F90                         <status=-1>
pe=00000 FAIL at line=01044    MAPL_Generic.F90                         <status=-1>
pe=00000 FAIL at line=01818    MAPL_Generic.F90                         <status=-1>
pe=00000 FAIL at line=01272    MAPL_Generic.F90                         <status=-1>
pe=00000 FAIL at line=01044    MAPL_Generic.F90                         <status=-1>
pe=00000 FAIL at line=01818    MAPL_Generic.F90                         <status=-1>
LANDInitialize                                1573
SURFACEInitialize                             3550
PHYSICS::Initialize                           1636
pe=00000 FAIL at line=01272    MAPL_Generic.F90                         <status=-1>
pe=00000 FAIL at line=01044    MAPL_Generic.F90                         <status=-1>
pe=00000 FAIL at line=01818    MAPL_Generic.F90                         <status=-1>
pe=00000 FAIL at line=01272    MAPL_Generic.F90                         <status=-1>
pe=00000 FAIL at line=01044    MAPL_Generic.F90                         <status=-1>
pe=00000 FAIL at line=01818    MAPL_Generic.F90                         <status=-1>
pe=00000 FAIL at line=00645    MAPL_CapGridComp.F90                     <status=-1>
pe=00000 FAIL at line=00949    MAPL_CapGridComp.F90                     <status=-1>
pe=00000 FAIL at line=00299    MAPL_Cap.F90                             <status=-1>
pe=00000 FAIL at line=00258    MAPL_Cap.F90                             <status=-1>
pe=00000 FAIL at line=00192    MAPL_Cap.F90                             <status=-1>
pe=00000 FAIL at line=00169    MAPL_Cap.F90                             <status=-1>
pe=00000 FAIL at line=00034    GEOSgcm.F90                              <status=-1>
AGCMInitialize                                1194
GCMInitialize                                 1233
--------------------------------------------------------------------------
MPI_ABORT was invoked on rank 0 in communicator MPI_COMM_WORLD
with errorcode 10928.

NOTE: invoking MPI_ABORT causes Open MPI to kill all MPI processes.
You may or may not see output from other processes, depending on
exactly when Open MPI kills them.
--------------------------------------------------------------------------
>> Error << /discover/swdev/gmao_SIteam/MPI/openmpi/4.1.3/gcc-12.1.0/bin/mpirun  -np 1 ./GEOSgcm.x : status = 176; at /gpfsm/dnb44/mathomp4/SystemTests/builds/AGCM_GNU/CURRENT/GEOSgcm/install-Debug/bin/esma_mpirun line 378.
```

It's almost like the *order* of the vegdyn restart changed...but it has barely been touched!

We might need the help of @gmao-rreichle and his group to figure this one out.

**Comment by mathomp4** [comment]
> @mathomp4, I took a brief look and have no clue how the changes in the branch ("Files Changed") could possibly be related to the VegDyn restart. As you said, nothing changed in VegDyn in a long time. Could this be an issue with the test setup?

Maybe. 🤷🏼 I mean, the SCM cases are nigh unchanging. But, the `vegdyn.data` files for the SCM cases haven't changed in a long time either (probably many many years). I assume they are good, but they are also 12-byte binary files. 

I might need to consult @bena-nasa or @atrayano on this. Binary restarts are an arcane science to me.

**Comment by gmao-rreichle** [comment]
FWIW, vegdyn.data should contain three variables in (land-only) tile space in a Fortran sequential binary file:

https://github.com/GEOS-ESM/GEOSgcm_GridComp/blob/68d5bbb718a94dce052abc1e4d5d799de20ddcd4/GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/Utils/Raster/makebcs/create_README.csh#L1039

 ```
        file name: vegdyn.data or ../vegdyn_*.dat
         file format: fortran binaries, little_endian
             read ([UNIT]) (primary_veg_type(n),n=1,${NTILES})
             read ([UNIT]) (canopy_height   (n),n=1,${NTILES})
             read ([UNIT]) (ASCATz0         (n),n=1,${NTILES})
```

The vegdyn.data files in the (many!) SCM test directories all have a file size 12 bytes, which made me wonder.  I then proceeded to see how many bytes I would expect and looked at the Icarus-NLv3 EASEv2 M36 bcs:

/gpfsm/dnb06/projects/p185/legacy_bcs/Icarus-NLv3/Icarus-NLv3_EASE/SMAP_EASEv2_M36/vegdyn_964x406_DE.dat

This particular tile space has 112,581 land (type=100) tiles.  With three fields and 4 bytes per data point this should make 3*4*112,581 = 1,350.876 bytes (not counting the 6 integers that should also be part of the Fortran sequential binary file).   But the above file size is 1,377,347 bytes.  I have no idea what's going on.  I'm sure I'm missing something.

**Comment by gmao-rreichle** [comment]
Well, the vegdyn file in NLv3 is nc4, which explains the file size discrepancy:
https://github.com/GEOS-ESM/GEOSgcm_GridComp/blob/68d5bbb718a94dce052abc1e4d5d799de20ddcd4/GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/Utils/Raster/makebcs/mod_process_hres_data.F90#L1045

I still have no clue what might have changed that the branch won't work with binary vegdyn restart data.  I think we could use GEOSldas setup to create single-tile ("zoomed-in") vegdyn restart files for the SCM test case(s), if that helps.  @weiyuan-jiang would know better.

**Comment by gmao-rreichle** [comment]
@mathomp4:  After poking around a bit more in your SCM test dir, I'm increasingly convinced that the vegdyn.data files there are Fortran sequential binary files.  That is, the first 4 bytes in the file represent the integer number of bytes of data (i.e., 4) to be read in the record, followed by the veg class in the next 4 bytes (represented as 4-byte float), and the final 4 bytes match the first four bytes in the record.  That is, this particular set of vegdyn.data files contains **only** the veg class.  The veg height and ASCAT z0 info is not present.  I don't remember what is required and what can be cold-started.  In older tags, veg height was from a look-up table and may not have been required in the vegdyn restart.  I suppose newer tags work only if veg height and ASCAT z0 are also provided in the restart file (despite the fact that ASCAT z0 is essentially garbage and never used...).   
In other words, I think the vegdyn.data files in the SCM tests are dated.
What confuses me is that the SCM tests seem to have been running ok until recently, or perhaps I misunderstood this.

**Comment by biljanaorescanin** [comment]
Just a small comment "vegdyn.data" file is binary only for old land boundary conditions, since NLv3 it was netCDF4. So what ever is used here is at least from 2019 or earlier. And one more update was done on it while file is nc4, in 2022 we fixed file size for nc4 from "unlimited" to max number of tiles. 
I will update "create_README.csh" to reflect this.

**Review by sdrabenh — APPROVED** [review]
